Rewritten TeamSpec::balance_extents() to balance more than two dimensions

ddiefenthaler opened the pull request:
I've rewritten balance_extents to use a different algorithm.

ddiefenthaler committed:
Added documentation to TeamSpec::balance_extents

ddiefenthaler committed:
Rewritten TeamSpec::balance_extents to work with higher dimensions

ddiefenthaler committed:
added example to documentation of TeamSpec::balance_extents

ddiefenthaler commented:
The idea behind the algorithm is to equally distribute the prime factors across the dimensions as good as possible. Starting with the largest factors they get multiplied with `this->extents[i]` with `i` growing.
As long as there are enough remaining prime factors this condition should be true:
`this->_extents[0] >= this->_extents[1] >= ... >= this->_extents[MaxDimensions-1]`
If the last iteration over `this->_extents` is incomplete the direction gets reversed to even the effect out. (Starting with `this->_extents[MaxDimensions-1]` downwards)

Thinking about it, it may make sense to change the direction each iteration over `this->_extents`,
but I'm not sure about this... I can't see what would happen after the second iteration.

fuchsto reviewed (changes requested):
Thank you, but please implement additional unit tests to cover the new functionality.

fuchsto commented:
Addresses #228

ddiefenthaler committed:
Added unit test for TeamSpec.balance_extents

ddiefenthaler committed:
Fixed an error in balance_extents

ddiefenthaler commented:
I added a unit test,
but I think that the algorithm should still be changed once more.
I assume that my approach only works well if the number of units factorizes in not to many different factors.
Apart from that the algorithm does not consider any constraints.
@fuchsto It would be nice if you could clarify how balance_extents should behave.

ddiefenthaler committed:
Changed the algorithm in balance_extents to balance more equally

ddiefenthaler committed:
Merge branch 'development' into feat-228-teamspec-balance-extents

ddiefenthaler committed:
Removed adjustment of _is_linear

ddiefenthaler commented:
I added another commit, I think this is in a mergable state now.

I am sure there are more optimizations possible, but for now this works too.

I removed the adjustment of `_is_linear`, that led to `MatrixTest.CopyRow` failing,
but I think this is another error so I will open a seperate issue for that.

ddiefenthaler commented:
I think this Pull request is fine now,
@fuchsto can you review this again?

fuchsto committed:
Merge branch 'development' into feat-228-teamspec-balance-extents

fmoessbauer commented on dash/include/dash/TeamSpec.h:
```diff
@@ -218,39 +219,56 @@ class TeamSpec :
+      DASH_LOG_TRACE("TeamSpec.balance_extents()",
+                     "factor:", it->first, "x", it->second);
+      for (auto i = 1; i < it->second + 1; ++i) {
+        new_extents.insert(it->first * *new_extents.begin());
```
You are not a fan of brackets, are you? ;)

fuchsto committed:
Merge branch 'development' into feat-228-teamspec-balance-extents

fuchsto commented:
@ddiefenthaler Looking good!

fuchsto commented:
@ddiefenthaler I don't care too much about efficiency of this implementation for now. Integer factorization is NP-hard, anyways.
This function is supposed to be called once for a reasonably large unit of work, as in domain decomposition for a 2048x2048 matrix-matrix multiplication. Even if it was inefficient, it would not notably affect the overall time to completion.
